### PR TITLE
CI: add pydocstyle to code_checks

### PR DESCRIPTION
**Pull request by simonjayhawkins**
http://www.pydocstyle.org/en/5.0.2/error_codes.html

### CODE REVIEW

**Comment by jreback** [comment]
looks ok to me. @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
How does this contrast / overlap with the validations we already do with numpydoc?

**Comment by simonjayhawkins** [comment]
many differences so will only give the main points below. can go into more fine-grained detail if required.

pydocstyle is a static checker so is not able to check dynamically created docstrings.

validate_docstrings script is only checking public api. pydocstyle is checking internal docstrings too.

pydocstyle and validate_docstrings have some overlapping but also exclusive error codes so having both is beneficial.

pydocstyle has a flake8 plugin so can see errors highlighted in ide so a benefit for contributors.

not added here, but error code 'D417 | Missing argument descriptions in the docstring' would add a lot more than consistent style. It would ensure contributors properly document their functions/methods. A lot of work to enable this globally.

**Comment by WillAyd** [comment]
lgtm. I think we can start with this and refine over time if conflicts with numpydoc arise

**Review by WillAyd — APPROVED** [review]

File: requirements-dev.txt [hunk 1]
```diff
@@ -14,6 +14,7 @@ flake8-rst>=0.6.0,<=0.7.0
 isort
 mypy==0.730
 pycodestyle
+pydocstyle
```
Comment (WillAyd) [inline requirements-dev.txt]: Can you also add to environment.yml?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: requirements-dev.txt (hunk 1 above)
Comment (simonjayhawkins) [reply requirements-dev.txt]: it has been?

**Comment by simonjayhawkins** [comment]
> lgtm. I think we can start with this and refine over time if conflicts with numpydoc arise

Ok. we could enable straight away and see how things pan out. so will take off draft status.

i suppose if issues do arise can always disable. The error codes included here are the simpler ones and it was not a lot of effort to make the changes across the codebase to comply.

we could then look to update the style guide once happy that having on the ci is smooth.

**Comment by jorisvandenbossche** [comment]
> error code 'D417 | Missing argument descriptions in the docstring' would add a lot more than consistent style

That's something that numpydoc checks as well? 
And actually, as far as I can see, pydocstyle cannot do this for numpy-style docstrings?

File: pandas/core/arrays/datetimes.py [hunk 2]
```diff
@@ -820,7 +819,7 @@ def tz_convert(self, tz):
         dtype = tz_to_dtype(tz)
         return self._simple_new(self.asi8, dtype=dtype, freq=self.freq)
 
-    def tz_localize(self, tz, ambiguous="raise", nonexistent="raise"):
+    def tz_localize(self, tz, ambiguous="raise", nonexistent="raise"):  # noqa
```
Comment (jorisvandenbossche) [inline pandas/core/arrays/datetimes.py]: Why is the noqa needed here?

File: pandas/core/tools/datetimes.py [hunk 3]
```diff
@@ -538,7 +538,7 @@ def to_datetime(
     infer_datetime_format=False,
     origin="unix",
     cache=True,
-):
+):  # noqa: D207
```
Comment (jorisvandenbossche) [inline pandas/core/tools/datetimes.py]: same question here?

Comment (simonjayhawkins) [reply pandas/core/tools/datetimes.py]: there looks to be a problem with the use of continuation. The next line is flagged as underindented (D207).

File: pandas/core/arrays/datetimes.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/datetimes.py]: another continuation line issue.

Comment (jorisvandenbossche) [reply pandas/core/arrays/datetimes.py]: Can we rather figure out what is wrong, or disable the indentation check?

File: pandas/core/tools/datetimes.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/core/tools/datetimes.py]: It's black that requires this, so I would maybe not include the D207 check then, if it gives conflicts with black.

**Comment by simonjayhawkins** [comment]
>That's something that numpydoc checks as well?
>And actually, as far as I can see, pydocstyle cannot do this for numpy-style docstrings?

can you elaborate?

the checks of the parameters aren't broken down into separate codes for the parameter violations and the checks are not so rigorous.

sample cli output of D417
```
$ pydocstyle pandas/io/formats/format.py  --select=D417
pandas/io/formats/format.py:968 in public method `to_html`:
        D417: Missing argument descriptions in the docstring (argument(s) buf, encoding are missing descriptions in 'to_html' docstring)
pandas/io/formats/format.py:1111 in public function `format_array`:
        D417: Missing argument descriptions in the docstring (argument(s) decimal, digits, float_format, formatter, justify, na_rep, space, values are missing descriptions in 'format_array' docstring)
pandas/io/formats/format.py:1507 in public function `format_percentiles`:
        D417: Missing argument descriptions in the docstring (argument(s) percentiles are missing descriptions in 'format_percentiles' docstring)
(pandas-dev) 
```

File: pandas/core/tools/datetimes.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/tools/datetimes.py]: disabling D207 on one or two functions is better than disabling on all, surely?

File: pandas/core/arrays/datetimes.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/datetimes.py]: probably need a fix upstream

File: pandas/core/tools/datetimes.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/tools/datetimes.py]: see https://github.com/PyCQA/pydocstyle/issues/234

File: pandas/core/arrays/datetimes.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/arrays/datetimes.py]: looks like an issue has already been raised for this https://github.com/PyCQA/pydocstyle/issues/437

File: ci/code_checks.sh [hunk 4]
```diff
@@ -316,6 +316,11 @@ fi
 ### DOCSTRINGS ###
 if [[ -z "$CHECK" || "$CHECK" == "docstrings" ]]; then
 
+
+    MSG='Validate docstrings (D201, D202, D204, D207, D208, D209, D213, D300, D409, D411, D412, D414)' ; echo $MSG
+    pydocstyle pandas --select=D201,D202,D204,D207,D208,D209,D213,D300,D409,D411,D412,D414
```
Comment (datapythonista) [inline ci/code_checks.sh]: Would it make sense to have the codes in `setup.cfg`? I guess we could then call `pydocstyle pandas` locally and get this validation.

**Review by datapythonista — APPROVED** [review]
lgtm

**Review by WillAyd — APPROVED** [review]
lgtm @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
@simonjayhawkins Thanks for the explanation above https://github.com/pandas-dev/pandas/pull/32033#issuecomment-586731312 So I suppose the main reason to use it in addition to numpydoc is that it checks all docstrings? 

> > And actually, as far as I can see, pydocstyle cannot do this for numpy-style docstrings?
>
> can you elaborate?

That was mentioned in one of the issues on numpydoc that talked about using pydocstyle, but it was from a while ago, so probably out of date

---

If this is going to fail CI in case of violation, we should add this to the contributing guidelines (similar to black and flake8)

Also, should we add it to pre-commit hooks? (can it run on only the changed files? Or how fast it is for the full code base?)

**Review by jorisvandenbossche — CHANGES_REQUESTED** [review]
So looks good, but needs some docs IMO

**Comment by alimcmaster1** [comment]
Maybe worth a note in the contributing guide?

**Comment by simonjayhawkins** [comment]
I seem incapable of successfully building documentation locally on my Windows machine. closing to clear queue

**Comment by jorisvandenbossche** [comment]
@simonjayhawkins for only adding some prose, there is not necessarily a need to build the docs locally. I can certainly review on technical rst aspects.

**Comment by simonjayhawkins** [comment]
I think the ideal (but not most straightforward approach) is to update https://dev.pandas.io/docs/development/contributing_docstring.html.

I think that we need to avoid confusion between the public API functions/methods/classes that use sphinx to generate documentation for pandas users and docstrings for all other (internal) functions/method/classes that are for the benefit of pandas contributors.

A simpler approach for now could be to add a one-liner in the contributing guide instead.

**Comment by WillAyd** [comment]
Do we have a go forward on this? I still think OK and would be fine with doc updates as a follow on since this isn't drastically changing things.

Though not a strong preference either way - just don't want this to hang in limbo if it doesn't need to

**Comment by jorisvandenbossche** [comment]
If it is going to fail CI, it should be documented in the contributing docs. It doesn't need to belong, just a short mention of it should be fine.

Also:

> Also, should we add it to pre-commit hooks? (can it run on only the changed files? Or how fast it is for the full code base?)

**Comment by simonjayhawkins** [comment]
> If it is going to fail CI, it should be documented in the contributing docs.

#32393 was intended as  precursor to this. Do we have the contributing docs available now? if so, can you provide a link, and i'll reopen #32393.

**Comment by simonjayhawkins** [comment]
> Also, should we add it to pre-commit hooks?

i don't have these (may require a bit of effort/knowledge to setup on Windows). so will need others to push the relevant patch for this.

**Comment by simonjayhawkins** [comment]
> Though not a strong preference either way - just don't want this to hang in limbo if it doesn't need to

no strong preference either. see https://github.com/pandas-dev/pandas/pull/32033#issuecomment-586744504.

**Comment by simonjayhawkins** [comment]
I think we need further discussion before we address https://github.com/pandas-dev/pandas/pull/32033#issuecomment-586727246 and then update the guidance. I don't think this PR is that vehicle.

**Comment by simonjayhawkins** [comment]
> I think we need further discussion before we address [#32033 (comment)](https://github.com/pandas-dev/pandas/pull/32033#issuecomment-586727246) and then update the guidance. I don't think this PR is that vehicle.

opened #32773, closing this to clear the queue for now.

**Pull request closed without merging**